pkg/resourcesets/collector.go

```diff
@@ -74,18 +73,16 @@ func (h *ResourceHandler) GatherResources(ctx context.Context, resourceSelectors
 				// if this is a subresource, check if its a status subsubresource
 				if split[1] == "status" && split[0] != "customresourcedefinitions" && slice.ContainsString(res.Verbs, "update") {
```

Comment:
Previously we have skipped (`continue`) even if this check didn't succeed, now we skip only if `if` is true.

I think we should remove this `if` altogether and change comment to state that we don't gather subresources separately (note `scaling` is also a subresource).

Reply:
As soon as checks below result in us skipping querying `crd-name/status` and `crd-name/scaling` subresources directly then sure (do they have `list` or `get` verbs available? - I'm not sure). But it may be cleaner to just leave this check (`if len(split) == 2 {`) so it's clear we're explicitly skipping subresources.

Reply:
I keep the original logic and add a debugging message.

Reply:
We don't really do anything on backup when we detect status subresource, so I think it's pointless to do this check or to log this particular debug message.
I would change the whole thing to something like this (you can add a comment to explain, but I think debug message text pretty much IS that comment):
```go
if string.Contains(res.Name, "/") {
	logrus.Debugf("Skipped backing up subresource: %s", res.Name)
	continue
}
```

Reply:
To be safe, I prefer to keep the existing check because `if string.Contains(res.Name, "/")` catches more cases than the existing logic although I am not sure if they exist.

Reply:
@jiaqiluo , the current logic uses `SplitN` which returns not more than N matches. So I believe `len(strings.SplitN(res.Name, "/", 2)) == 2` is equivalent to `strings.Contains(res.Name, "/")` - I cannot think of even edge cases where these would evaluate to different values.

Also note I really want to have this long check `split[1] == "status" && split[0] != "customresourcedefinitions" && slice.ContainsString(res.Verbs, "update")` removed - our new logic is simple - if the resource we're examining is a subresource, we're skipping it.

To sum up, the suggestion is to replace current code:
```go
			split := strings.SplitN(res.Name, "/", 2)
			if len(split) == 2 {
				// if this is a subresource, check if its a status subsubresource
				if split[1] == "status" && split[0] != "customresourcedefinitions" && slice.ContainsString(res.Verbs, "update") {
					// this resource has status subresource and it accepts "update" verb, so we need to call UpdateStatus on it during restore
					logrus.Debugf("Skipped backing up subresource: %s", res.Name)
				}
				// no need to save contents of any subresource as they are a part of the resource
				continue
			}
```

with something much shorter:
```go
			if strings.Contains(res.Name, "/") {
				logrus.Debugf("Skipped backing up subresource: %s", res.Name)
				continue
			}
```

Please let me know if I'm missing something in my logic.

Reply:
This long check `split[1] == "status" && split[0] != "customresourcedefinitions" && slice.ContainsString(res.Verbs, "update")` is what I was worried about since I wanted to avoid catching&skipping unexpected resources. 

But because it looks like there are only 2 possibilities which the res.Name containing `/`:
either `xxx/staus` or `xxx/scale`, I will take your suggested fix :) 

```
> cat test-logs.txt| grep res.Name
INFO[2021/08/23 15:41:21] res.Name: customresourcedefinitions/status
INFO[2021/08/23 15:41:22] res.Name: customresourcedefinitions/status
INFO[2021/08/23 15:41:23] res.Name: customresourcedefinitions/status
INFO[2021/08/23 15:41:23] res.Name: gitreporestrictions/status
INFO[2021/08/23 15:41:23] res.Name: clustergroups/status
INFO[2021/08/23 15:41:23] res.Name: imagescans/status
INFO[2021/08/23 15:41:23] res.Name: clusterregistrations/status
INFO[2021/08/23 15:41:24] res.Name: clusters/status
INFO[2021/08/23 15:41:24] res.Name: bundledeployments/status
INFO[2021/08/23 15:41:24] res.Name: clusterregistrationtokens/status
INFO[2021/08/23 15:41:24] res.Name: gitrepos/status
INFO[2021/08/23 15:41:24] res.Name: bundlenamespacemappings/status
INFO[2021/08/23 15:41:24] res.Name: bundles/status
INFO[2021/08/23 15:41:24] res.Name: gitjobs/status
INFO[2021/08/23 15:41:24] res.Name: customresourcedefinitions/status
INFO[2021/08/23 15:41:25] res.Name: customresourcedefinitions/status
INFO[2021/08/23 15:41:25] res.Name: clusters/status
INFO[2021/08/23 15:41:25] res.Name: vmwarevspheremachines/status
INFO[2021/08/23 15:41:26] res.Name: amazonec2machines/status
INFO[2021/08/23 15:41:26] res.Name: linodemachines/status
INFO[2021/08/23 15:41:26] res.Name: azuremachines/status
INFO[2021/08/23 15:41:26] res.Name: digitaloceanmachines/status
INFO[2021/08/23 15:41:26] res.Name: rkebootstraptemplates/status
INFO[2021/08/23 15:41:26] res.Name: rkebootstraps/status
INFO[2021/08/23 15:41:26] res.Name: custommachines/status
INFO[2021/08/23 15:41:26] res.Name: rkecontrolplanes/status
INFO[2021/08/23 15:41:26] res.Name: rkeclusters/status
INFO[2021/08/23 15:41:26] res.Name: machines/status
INFO[2021/08/23 15:41:26] res.Name: machinedeployments/status
INFO[2021/08/23 15:41:26] res.Name: machinedeployments/scale
INFO[2021/08/23 15:41:26] res.Name: clusters/status
INFO[2021/08/23 15:41:26] res.Name: machinesets/status
INFO[2021/08/23 15:41:26] res.Name: machinesets/scale
INFO[2021/08/23 15:41:26] res.Name: machinehealthchecks/status
INFO[2021/08/23 15:41:26] res.Name: projects/status
INFO[2021/08/23 15:41:27] res.Name: clusters/status
INFO[2021/08/23 15:41:27] res.Name: roletemplatebindings/status
INFO[2021/08/23 15:41:27] res.Name: roletemplates/status
INFO[2021/08/23 15:41:27] res.Name: customresourcedefinitions/status
INFO[2021/08/23 15:41:29] res.Name: customresourcedefinitions/status
INFO[2021/08/23 15:41:29] res.Name: managedcharts/status
INFO[2021/08/23 15:41:30] res.Name: apiservices/status
INFO[2021/08/23 15:41:32] res.Name: resourcesets/status
```

Reply:
@jiaqiluo , I think there is still some confusion and I would like us to be on the same page so you don't feel like I'm pushing my suggestion.

Current code skips resource as soon as there is a subresource, regardless of the result of that long `if` check. Long `if` determines whether we are adding this CRD to `statussubresource.json` and since we no longer need this, it's absolutely safe to remove this long check.

See https://github.com/rancher/backup-restore-operator/blob/e4e8c80e8f6300c42df60f9ff393132bd5069fea/pkg/resourcesets/collector.go#L72-L82, `continue` is within `if len(split) == 2` "then" block. And since we no longer need `split[i]`, my suggestion is to replace it with the simpler `Contains` check.

Are we on the same page now? I feel like I may be missing something, so really want you to confirm.

Reply:
Yep, we are are on the same page now 😆 

I really appreciate and like the conversation with you. Please keep doing it 💪